### ISSUE TITLE
Unexpected SIGTERM handling

### ISSUE THREAD

**Issue by qwelias** [pnpm/pnpm#2653]
### Code to reproduce the issue:
```js
// sigterm.js
process.on('SIGTERM', (...args) => {
    console.log(...args);
    setTimeout(process.exit, 5000, 123);
});
setInterval(console.log, 1000, 1);
```
```json
// package.json
...
"script": {
  "sigterm": "node sigterm.js"
}
...
```
1. `pnpm run sigterm`
2. in another tty `kill -TERM <pid_of_pnpm_run_sigterm>`

### Expected behavior:
`pnpm run sigterm` upon receiving SIGTERM should pass the signal to child and wait for it to exit, pass 123 exit code to parent.

### Actual behavior:
`pnpm run sigterm` upon receiving SIGTERM passes signal to child, but does not wait for it to exit, hence child's exit code is lost (and child is potentially running detached?).

### Additional information:
- npm handles it flawlessly
- [yarn has similar issue open](https://github.com/yarnpkg/yarn/issues/4667) for almost 3yrs now, don't be like yarn

EDIT:
- change 666 to 123 because it's capped at 154

**Comment by zkochan** [comment]
I am not sure why it works differently with pnpm as we use a fork of npm's library for running the scripts. And it has the following code:

```js
process.once('SIGINT', procInterupt)

  function procInterupt () {
    proc.kill('SIGINT')
    proc.on('exit', () => {
      process.exit()
    })
    process.once('SIGINT', procKill)
  }
```

https://github.com/zkochan/lifecycle/blob/latest/index.js

**Comment by zkochan** [comment]
oh, they probably use it only for running lifecycle scripts of dependencies during installation. Maybe `npm run` has some other implementation.

**Comment by qwelias** [comment]
@zkochan 
it seems like npm just kills itself upon `SIGTERM` according to [relative source file](https://github.com/npm/cli/blob/abdf52879fcf0e0f534ad977931f6935f5d1dce3/node_modules/npm-lifecycle/index.js#L343), but I'm not sure if that's the whole picture.

Here's an example output comparison of npm vs pnpm:
(The logs also include terminal prompt which shows last non-zero exit code after timestamp.)
<details>
  <summary>`npm run sigterm` gets SIGTERM</summary>

```
qwelias@schibmaze ~/prj/stuff  (1.0.0)                                                                                                                                    ⬡ 12.18.1 [10:48:20]
$ npm run sigterm  

> stuff@1.0.0 sigterm /home/qwelias/prj/stuff
> node sigterm

1
1
1
1
1
SIGTERM 15
1
1
1
1
1
npm ERR! code ELIFECYCLE
npm ERR! errno 123
npm ERR! stuff@1.0.0 sigterm: `node sigterm`
npm ERR! Exit status 123
npm ERR! 
npm ERR! Failed at the stuff@1.0.0 sigterm script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/qwelias/.npm/_logs/2020-06-25T08_48_37_781Z-debug.log

qwelias@schibmaze ~/prj/stuff  (1.0.0)                                                                                                                                ⬡ 12.18.1 [10:48:37] 123

```
npm exits only after it's child is exited, exit code passed.
</details>

<details>
  <summary>`pnpm run sigterm` gets SIGTERM</summary>

```
qwelias@schibmaze ~/prj/stuff  (1.0.0)                                                                                                                                ⬡ 12.18.1 [10:55:27]
$ pnpm run sigterm  

> stuff@1.0.0 sigterm /home/qwelias/prj/stuff
> node sigterm

1
1
1
1
1
1
SIGTERM 15

qwelias@schibmaze ~/prj/stuff  (1.0.0)                                                                                                                                    ⬡ 12.18.1 [10:55:46]
$ 1
1
1
1
1


qwelias@schibmaze ~/prj/stuff  (1.0.0)                                                                                                                                    ⬡ 12.18.1 [10:55:55]

```
pnpm exited immediately after forwarding `SIGTERM`, while the script continued running orphaned and attached to stdio, exit code was lost
</details>

**Comment by rhyek** [comment]
A general demonstration of the issue:

![image](https://user-images.githubusercontent.com/702922/146280170-33e07e58-c7f6-4bac-8ebe-9d6d6342111a.png)

**Comment by rhyek** [comment]
Hi, was just wondering if this was being looked at. I've reverted to using `npm run` on our project and pnpm for everything else. pnpm is perfectly fine for us in every other sense and I see new features being released frequently that seem like they tackle edge cases that I've never run into in 1.5 years of using this library. In my humble opinion, this anomalous handling of system signals should be of higher priority.

**Comment by zkochan** [comment]
I did add it to a priority list but I have too much to do. And it is pretty hard for me to concentrate now, since I am in Ukraine.

I am OK to accept a contribution that makes it work the same way as it works in npm.

**Comment by rhyek** [comment]
Hope you are safe.

**Comment by DylanCulfogienis** [comment]
I'd be happy to contribute a PR for this if someone can just point me vaguely in the direction of where this is handled in pnpm. Already see the link above for npm.

Our team's been slowed down by this one for about a month and we'll have some extra time in the coming month or so to address lower-priority stuff for our project, including resolving this issue. Happy to give back a little to a project that's been instrumental to the success of our own.

**Comment by zkochan** [comment]
The issue is probably in this package: https://github.com/pnpm/npm-lifecycle

**Comment by DylanCulfogienis** [comment]
Sweet, I'll take a look soon here.

**Comment by mark-omarov** [comment]
Hey @DylanCulfogienis , are you still working on it? If there's any progress - please share.

**Comment by DylanCulfogienis** [comment]
@mark-omarov Afraid I haven't gotten to it yet- This work ultimately has to fit around the high-priority work for this project for me to justify it to our client.

If someone else stumbles across this and wants to give it a try, you're welcome to!

That being said, I will likely have time to take a look at this in late November/December. We're starting to clear out the high priority stuff, and this will be one of the first things I work on after that's said and done.

**Comment by mark-omarov** [comment]
@DylanCulfogienis thanks for the update. I have tried to figure it out, no luck yet.

All I can tell so far is that on termination I see three events fired - `process.once('SIGTERM', procInterupt)`, `proc.on('close', ...)`, and `process.on('exit')` ([file](https://github.com/pnpm/npm-lifecycle/blob/main/index.js)). For some reason, the process doesn't wait for its child indeed.
I've tried to ignore all except `SIGINT`, but the result was the same.

It's getting late, but I will take another look sometime later, and give an update if I figure something out.

P.S.
Also, the same behavior can be found with `pnpm exec`, but that works differently, there's no use of the mentioned `lifecycle`, I will try to find some time to check that out as well.

**Comment by paescuj** [comment]
It seems like the issue described in the OP has been solved in the meantime (at least on macOS, can others confirm as well?):
```console
❯ pnpm run sigterm

> monorepo-root@ sigterm <redacted path>
> node sigterm.js

1
1
1
SIGTERM 15
1
1
1
1
1

 ELIFECYCLE  Command failed with exit code 123.
```
(compare with output in https://github.com/pnpm/pnpm/issues/2653#issuecomment-649404664)

**Comment by zkochan** [comment]
Yes, it was fixed in v7.18.0

**Comment by ivstiv** [comment]
I am experiencing the broken behaviour that is said to be fixed. Just tried the OP's code to reproduce it and I ended up with a zombie process no matter whether I used pnpm, yarn or npm. Can anybody else confirm this or am I doing something stupid?

Versions used:
npm@9.4.2
pnpm@7.27.0

**Comment by qwelias** [comment]
@ivstiv 
npm@9.3.1
pnpm@7.27.0

works fine, but keep in mind this may also depend on your OS

**Comment by paymog** [comment]
I'm still seeing this issue on pnpm 8.10.0

**Comment by benweint** [comment]
Since this issue is marked as closed, I filed https://github.com/pnpm/pnpm/issues/7374 as a new issue describing the problem that @gajus described above.
